src/librustc_mir/interpret/cast.rs

```diff
@@ -239,17 +239,30 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpretCx<'mir, 'tcx, M> {
     fn cast_from_ptr(
         &self,
         ptr: Pointer<M::PointerTag>,
-        ty: Ty<'tcx>
+        src_layout: TyLayout<'tcx>,
+        dest_layout: TyLayout<'tcx>,
     ) -> InterpResult<'tcx, Scalar<M::PointerTag>> {
         use rustc::ty::TyKind::*;
-        match ty.sty {
+
+        match dest_layout.ty.sty {
             // Casting to a reference or fn pointer is not permitted by rustc,
             // no need to support it here.
-            RawPtr(_) |
-            Int(IntTy::Isize) |
-            Uint(UintTy::Usize) => Ok(ptr.into()),
-            Int(_) | Uint(_) => err!(ReadPointerAsBytes),
-            _ => err!(Unimplemented(format!("ptr to {:?} cast", ty))),
+            RawPtr(_) => Ok(ptr.into()),
+            Int(IntTy::Isize) | Uint(UintTy::Usize) => {
+                if let Ok(bits) = self.force_bits(Scalar::Ptr(ptr), self.memory.pointer_size()) {
+                    self.cast_from_int(bits, src_layout, dest_layout)
+                } else {
+                    Ok(ptr.into())
+                }
+            }
+            Int(_) | Uint(_) => {
+                if let Ok(bits) = self.force_bits(Scalar::Ptr(ptr), self.memory.pointer_size()) {
+                    self.cast_from_int(bits, src_layout, dest_layout)
```

Comment:
You can share this first part with the branch above.
Then in the `else`, you can do something like `if dest_layout.size == memory.pointer_size() { Ok(ptr.into()) } else { ... }`.

Reply:
Oh and actually, you should just forward the error, not make up a new one.
```rust
match self.force_bits(...) {
Ok(bits) => ...
Err(_) if dest_layout.size == memory.pointer_size() => Ok(ptr.into()), // continue with ptr value at int type
Err(e) => Err(e), // propagate error
}
```

Reply:
Even if the error I'm making is the one that was used in the original code?

Reply:
Better not to make assumptions like that. The intention here is to "catch the error if we can do better" (that's the `Ok(ptr)` branch). If we can't to better, no reason to catch and re-throw a new one.